src/jsglue.cpp

```diff
@@ -332,6 +332,30 @@ class WrapperProxyHandler : public js::Wrapper
     }
 };
 
+//http://searchfox.org/mozilla-central/source/js/public/Principals.h#24
+class ServoJSPrincipal : public JSPrincipals
+{
+    const void* origin; //box with origin in it
+  public:
+    ServoJSPrincipal(const void* origin)
```

Comment:
We will need to accept two function pointers as arguments here - one for `destroy` (that accepts a `*mut JSPrincipals` argument), and one for `write`. This will allow Rust code to provide callbacks that those C++ methods can invoke.